src/vm/instruction.ts

```diff
@@ -10,68 +33,303 @@
 // │                     off_op0 (biased representation)                     │
 // ├─────────────────────────────────────────────────────────────────────────┤
 // │                     off_op1 (biased representation)                     │
-// ├─────┬─────┬───────┬───────┬───────────┬────────┬───────────────────┬────┤
-// │ dst │ op0 │  op1  │  res  │    pc     │   ap   │      opcode       │ 0  │
-// │ reg │ reg │  src  │ logic │  update   │ update │                   │    │
-// ├─────┼─────┼───┬───┼───┬───┼───┬───┬───┼───┬────┼────┬────┬────┬────┼────┤
+// ├─────┬─────┬───────────┬───────┬───────────┬─────────┬──────────────┬────┤
+// │ dst │ op0 │    op1    │  res  │    pc     │   ap    │    opcode    │ 0  │
+// │ reg │ reg │    src    │ logic │  update   │ update  │              │    │
+// ├─────┼─────┼───┬───┬───┼───┬───┼───┬───┬───┼────┬────┼────┬────┬────┼────┤
 // │  0  │  1  │ 2 │ 3 │ 4 │ 5 │ 6 │ 7 │ 8 │ 9 │ 10 │ 11 │ 12 │ 13 │ 14 │ 15 │
 // └─────┴─────┴───┴───┴───┴───┴───┴───┴───┴───┴────┴────┴────┴────┴────┴────┘
 // Source: https://github.com/lambdaclass/cairo-vm_in_go/blob/main/pkg/vm/instruction.go
 
 // Dst & Op0 register flags
 // If the flag == 0, then the offset will point to Ap
 // If the flag == 1, then the offset will point to Fp
-enum RegisterFlag {
+export enum RegisterFlag {
   ApRegisterFlag = 0,
   FpRegisterFlag = 1,
 }
-type DstRegister = RegisterFlag;
-type Op0Register = RegisterFlag;
 
 // Op1Src
-enum Op1Src {
+export enum Op1Src {
   Imm = 0,
   FpPlusOffOp1 = 1,
   ApPlusOffOp1 = 2,
   Op0 = 4,
 }
 
 // ResLogic
-enum ResLogic {
+export enum ResLogic {
   Op1 = 0,
   Add = 1,
   Mul = 2,
   Unconstrained = 3,
 }
 
 // Pc Update
-enum PcUpdate {
+export enum PcUpdate {
   PcUpdateRegular = 0,
   PcUpdateJump = 1,
   PcUpdateJumpRel = 2,
   PcUpdateJnz = 3,
 }
 
 // Ap update
-enum ApUpdate {
+export enum ApUpdate {
   ApUpdateRegular = 0,
   ApUpdateAdd = 1,
-  AppUpdateAdd1 = 2,
-  AppUpdateAdd2 = 3,
+  ApUpdateAdd1 = 2,
+  ApUpdateAdd2 = 3,
 }
 
 // Fp Update
-enum FpUpdate {
+export enum FpUpdate {
   FpUpdateRegular = 0,
   FpUpdateApPlus2 = 1,
   FpUpdateDst = 2,
 }
 
-enum Opcode {
+export enum Opcode {
   NoOp = 0,
-  AssertEq = 1,
-  Call = 2,
-  Ret = 4,
+  Call = 1,
+  Ret = 2,
+  AssertEq = 4,
 }
 
-export type Instruction = {};
+export type Instruction = {
+  off0: Int16;
+  off1: Int16;
+  off2: Int16;
+  dstReg: RegisterFlag;
+  op0Reg: RegisterFlag;
+  op1Addr: Op1Src;
+  resLogic: ResLogic;
+  pcUpdate: PcUpdate;
+  apUpdate: ApUpdate;
+  fpUpdate: FpUpdate;
+  opcode: Opcode;
+};
+
+export function decodeInstruction(
+  encodedInstruction: Uint64
+): Result<Instruction, VMError> {
+  const highBit = 1n << 63n;
```

Comment:
I think every single one of these mask, offset, etc. needs to be commented with thorough and pedagogical explanation

Reply:
will add comments for them